openhands/agenthub/codeact_agent/function_calling.py

```diff
@@ -0,0 +1,311 @@
+```
+
+**Example 3: edit for long files**
+
+Given an existing file `/path/to/file.py` that looks like this:
+(1000 more lines above)
+1001|class MyClass:
+1002|    def __init__(self):
+1003|        self.x = 1
+1004|        self.y = 2
+1005|        self.z = 3
+1006|
+1007|print(MyClass().z)
+1008|print(MyClass().x)
+(2000 more lines below)
+
+The assistant wants to edit the file to look like this:
+
+(1000 more lines above)
+1001|class MyClass:
+1002|    def __init__(self):
+1003|        self.x = 1
+1004|        self.y = 2
+1005|
+1006|print(MyClass().y)
+(2000 more lines below)
+
+The assistant may produce an edit action like this:
+path="/path/to/file.txt" start=1001 end=1008
+content=```
+class MyClass:
+    def __init__(self):
+        # no changes before
+        self.y = 2
+        # self.z is removed
+
+# MyClass().z is removed
+print(MyClass().y)
+```
+"""
+
+FileEditTool = ChatCompletionToolParam(
+    type='function',
+    function=ChatCompletionToolParamFunctionChunk(
+        name='edit_file',
+        description=_FILE_EDIT_DESCRIPTION,
+        parameters={
+            'type': 'object',
+            'properties': {
+                'thought': {
+                    'type': 'string',
+                    'description': 'Reasoning about the file edit action.',
+                },
+                'path': {
+                    'type': 'string',
+                    'description': 'The absolute path to the file to be edited.',
+                },
+                'new_content_draft': {
+                    'type': 'string',
+                    'description': 'A draft of the new content for the file being edited. Note that the assistant may skip unchanged lines.',
+                },
+                'start': {
+                    'type': 'integer',
+                    'description': 'The starting line number for the edit (1-indexed, inclusive). Default is 1.',
+                },
+                'end': {
+                    'type': 'integer',
+                    'description': 'The ending line number for the edit (1-indexed, inclusive). Default is -1 (end of file).',
+                },
+            },
+            'required': ['path', 'content'],
+        },
+    ),
+)
+
+_MESSAGE_DESCRIPTION = """Send a message to the user.
+It is used for, but not limited to:
+* ask for clarification or provide additional information.
+* inform the user that the assistant have completed the task.
+"""
+
+MessageTool = ChatCompletionToolParam(
+    type='function',
+    function=ChatCompletionToolParamFunctionChunk(
+        name='message_user',
+        description=_MESSAGE_DESCRIPTION,
+        parameters={
+            'type': 'object',
+            'properties': {
+                'content': {
+                    'type': 'string',
+                    'description': 'The message content to send to the user. Ideally formatted in markdown.',
+                },
+            },
+            'required': ['message'],
+        },
+    ),
+)
```

Comment:
Oh, do we need this one at all? The MessageAction is default anyway, for CodeAct, when the LLM isn't trying to _do something_. Getting a regular response like some text just works, doesn't it?

Reply:
Since we are using function calling anyway - i figured maybe it will be easier if we just fix the action space for the agent AND force LLM to choose between one of them. But i guess what you said also make sense: maybe the model is trained better when responding to assistant message 🤔 maybe I'll revert

Reply:
yeah reverted - maybe this is closer to the training distribution. Thanks for pointing out!

Reply:
Revert back after 4 days -- for weaker model like Deepseek, if you don't force it to generate tool call via `tool_choice='required'` - it will be hard to make it generate tool calls 😓 

So i guess the easy way would be make ALL its action space as tool call and force a tool call for *each* action.